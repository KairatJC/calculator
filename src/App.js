import React from "react";
import CalcContext from "./CalcContext";
import Display from "./components/Display/Display";
import Keyboard from "./components/Keyboard/Keyboard";

class App extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      oldValue: null,
      operator: null,
      currentValue: "Set Value",
      isChanged: false,
      isDecimal: false
    };
  }

  render() {
    return (
      <CalcContext.Provider
        value={{
          oldValue: this.state.oldValue,
          currentValue: this.state.currentValue,
          operator: this.state.operator,
          insertValue: this.buttonHandler
        }}
      >
        <div className="App">
          <Display />
          <Keyboard />
        </div>
      </CalcContext.Provider>
    );
  }

  buttonHandler = value => {
    if (value === ".") {
      if (this.state.isDecimal) {
        return;
      } else {
        this.setState(state => ({
          isDecimal: true
        }));
      }
    }

    if (this.isNumber(value) || value === ".") {
      this.setState(state => ({
        currentValue: state.isChanged ? state.currentValue + value : value,
        isChanged: true
      }));
    }

    if (value === "c") {
      this.resetState();
    }

    if (this.isOperator(value) && this.state.isChanged) {
      this.makeOperation(value);
    }

    if (value === "=" && this.state.isChanged) {
      this.getResult();
    }
  };

  isNumber(str) {
    return /\d/.test(str);
  }

  isOperator(str) {
    return /[/*+-]/.test(str);
  }

  resetState() {
    this.setState(state => ({
      oldValue: null,
      operator: null,
      currentValue: "Set Value",
      isChanged: false,
      isDecimal: false
    }));
  }

  makeOperation(operator) {
    if (this.state.oldValue !== null) {
      this.getResult();
    }
    this.setState(state => ({
      oldValue: state.currentValue,
      operator: operator,
      currentValue: "",
      isDecimal: false
    }));
  }

  getResult() {
    if (this.state.currentValue === "") return;

    const oldValue = parseFloat(this.state.oldValue);
    const currentValue = parseFloat(this.state.currentValue);
    const operator = this.state.operator;
    this.resetState();
    this.setState(state => ({
      currentValue: this.calculate(oldValue, operator, currentValue)
    }));
  }

  calculate(operand1, operator, operand2) {
    if (operator === "+") {
      return operand1 + operand2;
    } else if (operator === "-") {
      return operand1 - operand2;
    } else if (operator === "*") {
      return operand1 * operand2;
    } else if (operator === "/" && operand2 !== 0) {
      return operand1 / operand2;
    }
  }
}

export default App;
