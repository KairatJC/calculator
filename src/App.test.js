import React from "react";
import renderer from "react-test-renderer";
import App from "./App";

test("Main App should render correctly", () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
