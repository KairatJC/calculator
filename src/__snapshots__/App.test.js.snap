// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main App should render correctly 1`] = `
<div
  className="App"
>
  <div
    className="calc wrap"
  >
    <div
      className="calc-display"
    >
      <span>
        Set Value
      </span>
    </div>
  </div>
  <div
    className="keyboard wrap"
  >
    <div
      className="numpad"
    >
      <button
        id="button-7"
        onClick={[Function]}
      >
        7
      </button>
      <button
        id="button-8"
        onClick={[Function]}
      >
        8
      </button>
      <button
        id="button-9"
        onClick={[Function]}
      >
        9
      </button>
      <button
        id="button-4"
        onClick={[Function]}
      >
        4
      </button>
      <button
        id="button-5"
        onClick={[Function]}
      >
        5
      </button>
      <button
        id="button-6"
        onClick={[Function]}
      >
        6
      </button>
      <button
        id="button-1"
        onClick={[Function]}
      >
        1
      </button>
      <button
        id="button-2"
        onClick={[Function]}
      >
        2
      </button>
      <button
        id="button-3"
        onClick={[Function]}
      >
        3
      </button>
      <button
        id="button-0"
        onClick={[Function]}
      >
        0
      </button>
      <button
        id="button-."
        onClick={[Function]}
      >
        .
      </button>
    </div>
    <div
      className="actionpad"
    >
      <button
        id="button-c"
        onClick={[Function]}
      >
        c
      </button>
      <button
        id="button-+"
        onClick={[Function]}
      >
        +
      </button>
      <button
        id="button--"
        onClick={[Function]}
      >
        -
      </button>
      <button
        id="button-*"
        onClick={[Function]}
      >
        *
      </button>
      <button
        id="button-/"
        onClick={[Function]}
      >
        /
      </button>
      <button
        id="button-="
        onClick={[Function]}
      >
        =
      </button>
    </div>
  </div>
</div>
`;
